design-proposals/nic-hotplug/nic-hotplug.md

```diff
@@ -0,0 +1,407 @@
+}
+```
+
+### VMIS
+```golang
+type VirtualMachineInstanceNetworkInterface struct {
+...
+    // If the interface is hotplugged, this will contain its status
+    HotplugStatus *InterfaceHotplugStatus `json:"hotplugStatus,omitempty"`
+}
+
+type InterfaceHotplugStatus struct {
+    // Phase are specific phases for the hotplug volume.
+    Phase InterfaceHotplugPhase `json:"phase,omitempty"`
+
+    // DetailedMessage has more information in case of failed operations.
+    DetailedMessage string `json:"DetailedMessage,omitempty"`
+}
+
+type InterfaceHotplugPhase string
+
+const InterfaceHotplugPhasePending    InterfaceHotplugPhase = "Pending"
+const InterfaceHotplugPhaseInfraReady InterfaceHotplugPhase = "InfraReady"
+const InterfaceHotplugPhaseReady      InterfaceHotplugPhase = "Ready"
+const InterfaceHotplugPhaseFailed     InterfaceHotplugPhase = "Failed"
+```
+
+
+### Hotplug for pods
+Assuming the following two `network-attachment-definition`s:
+```yaml
+---
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: macvlan-conf-1
+spec:
+  config: '{
+            "cniVersion": "0.3.0",
+            "type": "macvlan",
+            "master": "eth1",
+            "mode": "bridge",
+            "ipam": {
+                "type": "host-local",
+                "ranges": [
+                    [ {
+                         "subnet": "10.10.0.0/16",
+                         "rangeStart": "10.10.1.20",
+                         "rangeEnd": "10.10.3.50",
+                         "gateway": "10.10.0.254"
+                    } ]
+                ]
+            }
+        }'
+---
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: macvlan-conf-2
+spec:
+  config: '{
+            "cniVersion": "0.3.0",
+            "type": "macvlan",
+            "master": "eth1",
+            "mode": "bridge",
+            "ipam": {
+                "type": "host-local",
+                "ranges": [
+                    [ {
+                         "subnet": "12.10.0.0/16",
+                         "rangeStart": "12.10.1.20",
+                         "rangeEnd": "12.10.3.50"
+                    } ]
+                ]
+            }
+        }'
+```
+
+And a pod with the following spec:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod-case-03
+  annotations:
+    k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+spec:
+  containers:
+  - name: pod-case-03
+    image: docker.io/centos/tools:latest
+    command:
+    - /sbin/init
+```
+Update the pod to:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod-case-03
+  annotations:
+    k8s.v1.cni.cncf.io/networks: macvlan-conf-1,macvlan-conf-2
+spec:
+  containers:
+  - name: pod-case-03
+    image: docker.io/centos/tools:latest
+    command:
+    - /sbin/init
+```
+
+The aforementioned update will trigger multus to start the CNI ADD flow for the
+network named `macvlan-conf-2`.
+
+### Hotplug for VMs
+TODO
+
+## Functional Testing Approach
```

Comment:
Could you please add Pods test case in testing for regression?

Reply:
Done.